### PR TITLE
G20-498: Language support i18n

### PR DESCRIPTION
**Pull request by mpereiraesaa**

### CODE REVIEW

File: gulpfile.js [hunk 1]
```diff
@@ -33,13 +37,62 @@ const FONTS_DIR = 'build/fonts';
...
+    if (error !== null) {
+      process.stdout.write(error);
+    }
+  });
+}
+
+function fetchCaptions(data) {
+  const file = data.prodFile;
+  const promises = [];
+  const json = {};
```
Comment (asgeir-inspera) [inline gulpfile.js]: JSON is a format for string representation of objects. Avoid naming object variables "json".

File: gulpfile.js [hunk 2]
```diff
@@ -33,13 +37,62 @@ const FONTS_DIR = 'build/fonts';
...
+    mkdirp('i18n');
+  }
+  for (let i = 0; i < data.locales.length; i++) {
+    const locale = data.locales[i];
+    promises.push(fetchFile(webTranslateFile(webtranslateitReadKey, file, locale)));
+    json[locale] = {};
+  }
+  Promise.all(promises)
+  .then((responses) => {
+    Object.keys(json).forEach((key, index) => {
```
Comment (asgeir-inspera) [inline gulpfile.js]: Object.keys(json) does not guarantee any particular order of the returned keys.

File: gulpfile.js [hunk 3]
```diff
@@ -33,13 +37,62 @@ const FONTS_DIR = 'build/fonts';
...
+function fetchCaptions(data) {
+  const file = data.prodFile;
+  const promises = [];
+  const json = {};
+  if (!fs.existsSync('i18n')) {
+    mkdirp('i18n');
+  }
+  for (let i = 0; i < data.locales.length; i++) {
+    const locale = data.locales[i];
+    promises.push(fetchFile(webTranslateFile(webtranslateitReadKey, file, locale)));
```
Comment (asgeir-inspera) [inline gulpfile.js]: My suggestion:
```suggestion
    function fetchWebTranslateFile(webtranslateitReadKey, file, locale) {
        return fetchFile(webTranslateFile(webtranslateitReadKey, file, locale)).then((captions) => ({ locale, captions }));
    }
```

Then you do not need the Object.keys(json) looping below.

**Comment by jarlep** [comment]
Would be good if you actually try to use some captions too...

File: gulpfile.js (hunk 3 above)
Comment (mpereiraesaa) [reply gulpfile.js]: always improving things a lot, going to take note on your use of reducer method, thanks a lot

**Comment by mpereiraesaa** [comment]
> Would be good if you actually try to use some captions too...

I did it with a sample but later removed it because it was a POC. Going to use it today implementing G20-497 caption

File: src/sidebar/services/i18nService.js [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    11: 'es-CO',
+    12: 'pl-PL',
+  };
+  const supportedLanguages = {
+  'en-US': 'English',
+  'nb-NO': 'Norsk - bokmål',
+  'nn-NO': 'Norsk - nynorsk',
+  'pl-PL': 'Polski',
+  'sv-SE': 'Svenska',
+  'es-CO': 'Español de América Latina',
```
Comment (asgeir-inspera) [inline src/sidebar/services/i18nService.js]: Highly inconsistent indentation here.

File: src/sidebar/services/i18nService.js [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    });
+  }
+
+  /**
+   * Changes the language used for translations.
+   *
+   * @param {string} language IETF language code, eg: en-US or nn-NO
+   * @param {function} callback
+  */
+  function changeLanguage(language, callback) {
```
[the remark this answers is not in the document]
Comment (asgeir-inspera) [reply src/sidebar/services/i18nService.js]: I agree.

[the remark this answers is not in the document]
Comment (mpereiraesaa) [reply src/sidebar/services/i18nService.js]: done

**Comment by jarlep** [comment]
Good. We should also look at using env variable for webtranslateitReadKey.

File: gulpfile.js [hunk 6]
```diff
@@ -33,13 +37,70 @@ const FONTS_DIR = 'build/fonts';
 const IMAGES_DIR = 'build/images';
 const TEMPLATES_DIR = 'src/sidebar/templates';
 
+const webtranslateit = 'https://webtranslateit.com/api/projects/';
+const webtranslateitReadKey = process.env.WEBTRANSLATEITREADKEY || 'proj_pub_MpJb4ubUp8fjmDq6cwyQBw';
```
Comment (asgeir-inspera) [inline gulpfile.js]: This still reveals the key.

Comment (jarlep) [inline gulpfile.js]: Please also remove the key. The point is that it should be a bit more obscure to find it (even if it now is in git history)
Add a little note in https://inspera.atlassian.net/wiki/spaces/TESTC/pages/429621250/Grading+Hypothesis on how to use it. We should only use the env variable, and I will update the build server.

Comment (mpereiraesaa) [reply gulpfile.js]: removed.

File: README.md [hunk 7]
```diff
@@ -1,6 +1,13 @@
+Grading
+================
+
+You need to set the WEBTRANSLATE.IT READ KEY for the captions project based on hypothesis-client as an environment variable.
+
 Hypothesis client
 =================
 
+|| 'proj_pub_MpJb4ubUp8fjmDq6cwyQBw'
```
Comment (asgeir-inspera) [inline README.md]: Putting the key here will not make it any more secret. We should just document it in Confluence for now.

**Pull request merged**